working on session panel, corollary fixes

jvanasco opened the pull request:
I ran into some issues when working on an update to my package pyramid_session_multi, and decided to finally tackle #313 and backport some stuff into the standard debugtoolbar.

1. This introduces a new Session Panel to visualize the changes to a session. There is a Session section on the RequestVars panel; either that could be removed or this could go into that panel.  

2. One of my tests inserted  a float via `random.random()` into the session. This caused a fatal exception under Python3 due to sorting comparisons between a float and string.

  * I updated several `sorted` methods in `pyramid_debugtoolbar` to handle this
  * When the RequestVars and/or Headers panel(s) failed from the sort, the Headers panel never received a `.response` attribute and created another fatal Exception, which obfuscated/buried the stack trace about the original exception.  I adjusted that panel with a default `.response` attribute and a conditional check.

Tests are provided, along with an example of the Py3 sorting exception; I figured I would include that with an AssertRaises, in case that behavior changes and the workarounds are no longer needed.

This works as intended, but I am considering it a WIP that needs feedback for two reasons:

* There is some duplication of data between two panels
* The RequestVars panel does a fancy trick to access the session via the `wrap_load`. I'm not sure if that is compatible or necessary with my attempt at this panel, as the new panel must always access the session.

Because of this, I think the right approach may to be keep RequestVars as-is, introduce the new panel, but have it controlled on a per-request/cookie basis like the profiling panel.

mmerickel commented:
I'm hesitant to consider adding any panel that is off by default. Even the performance panel shows some basic information when disabled. I'd expect this to use the lazy load features to avoid loading / accessing the session unless it was loaded by the request itself.

mmerickel commented:
That being said, the idea of a session panel being included is nice. I think we should add one. Could you include a screenshot of what this panel looks like? Obviously if we're merging this we'll want one anyway included in the docs, but you don't have to go that far yet.

jvanasco commented:
Perhaps the "panel" could lazy-load by default; but would always show the information if a cookie/request instructed it to.   (panel in quotes, because this could be wrapped into the RequestVars panel)

Here's a quick screencap from a test environment. The blue exclamation badges mark that value has changed. The keys are sorted, and we see the "in" and "out" values. 

![Screen Shot 2020-11-02 at 10 49 23 AM](https://user-images.githubusercontent.com/204779/97888518-3796db80-1cf9-11eb-98ca-69d5a35145e5.png)

mmerickel commented:
If you feel that strongly that you should be able to trigger session loading then fine, there can be some sort of toggle for it, but I'm not seeing it being useful. I want to see session data on requests that need it, and don't want to trigger all the side-effects that come with loading a session on requests that don't need it. It makes the behavior of the app fairly significantly different than when the toolbar is disabled in production.

jvanasco commented:
I definitely understand and appreciate that, and don't want to change that functionality - which is why i brought up the "always on" setting to start with.

Also, it does seem like a major deficiency in this PR is not noting if the session was accessed.

For context:

1. Inherent to the functionality of this panel, there is only one opportunity to read the session data "in" - which is why it's "always on"*.  The "smart" features you describe can only happen on the session data "out".  (*The original values of cookie based sessions could be copied and analyzed, but that doesn't work for server-side sessions)

2. I've often used the "always show the session info" to debug situations where the necessary info to determine the problematic user-path was in the session.

jvanasco commented:
I stripped tests because this is still WIP, but I wanted to see if this current approach seems okay.

* By default, the panel only displays session data that was loaded during the request.
* The checkbox can be used to enable an advanced mode, where it ALWAYS loads the session, both before and after the request processing.
* the request_vars wrapped load approach is used to track if the session was accessed during the main request or not. this is showcased to the user.

While doing this, I also found and corrected a small bug in the toolbar javascript.  the string join and splits for active panels were on different characters, and duplication was possible.

jvanasco commented:
converted this to draft; still working on tests.

mmerickel commented on src/pyramid_debugtoolbar/static/toolbar/toolbar.js:
```diff
@@ -28,12 +28,13 @@ $(function() {
 
 
   $('#settings .switchable').click(function() {
+  	debugger;
```
err

jvanasco replied:
Whoops!

mmerickel reviewed (changes requested):
I'm not clear on the disconnect here but the `wrap_load` function from the `request_vars` panel will allow you to capture the ingress values. There is no need for a setting. This function invokes a callback on your class when it is first loaded, where you can copy the values. Then later you can process the response, check if the values are still the same, etc. You should be using that function in the same way that it is working in `request_vars`. If it isn't crystal clear, the `request_vars` panel is capturing the INGRESS values of variables like the session already. And only when it's accessed inside the request.

Finally, the session should be removed from the request vars panel entirely if it's being moved into this new panel.

mmerickel reviewed (changes requested):
You should also use `pyramid_debugtoolbar.utils.dictrepr`, similarly to the request vars panel.

jvanasco commented:
The current attempt - I converted to a draft while I iron some stuff out - is to implement the wrap_load from the request panel as default. That will offer the same exact functionality as the session panel, as you described above.

The setting will enable a new functionality, which will load the session data immediately.  It uses a modified version of the wrap load, which will alert the panel if the session was accessed.

So, by default: the same functionality. The ingress values are only logged when the session is accessed by a user.  But if the new setting is enabled, the ingress values are always logged on panel setup - so they are available when trying to debug problematic user flows.

mmerickel commented:
I’m really having trouble buying what you’re selling. If the request doesn’t use the session then I really don’t think the toolbar should load it.

mmerickel commented on tests/test_panels/test_session.py:
```diff
@@ -0,0 +1,390 @@
+    from pyramid.session import SignedCookieSessionFactory as Sessn
+except ImportError:
+    # pyramid_debugtoolbar supports Pyramid==1.4
+    from pyramid.session import UnencryptedCookieSessionFactoryConfig as Sessn
```
I'd be fine bumping the minimum supported version if it helps.

mmerickel replied:
However, doing that might actually make more work for you as we have to go through and remove dead code. So maybe this is better for now.

jvanasco commented:
How does this version look? The docs and screenshot have been updated. (see https://github.com/jvanasco/pyramid_debugtoolbar/blob/feature-session_pre_post/docs/session.png )

The minimum supported version isn't an issue for me. I'm glad it works on earlier Pyramid.

I redid the wrapped loaders.  By default the package only tracks ingress/egress on accessed sessions.  Using a cookie, either per-request or via the toolbar settings, will enable the toolbar to peek inside the session for advanced debugging.  In both circumstances, the toolbar informs developers if the Session has been accessed or not.

I updated the `toolbar.js` to use a variable as the delimiter. this was only done to aggregate the cookie name and panel delimiter at the top of the javascript file, so they could be documented alongside each other - and referenced back to `toolbar.py` in comments.  the python file also references the javascript file now too.  I added comments in a few places were things weren't too clear.

The "peek inside the session" functionality has saved us countless hours on debugging complex edge cases.  Even if the session isn't accessed, it lets us look at the internal structure - such as user ids - so problematic user flows can be detected. If you are using browser tools to swap between cookie profiles, enabling this lets you quickly document the exact user flow.

stevepiercy commented on src/pyramid_debugtoolbar/panels/templates/session.dbtmako:
```diff
@@ -0,0 +1,90 @@
+<h5>Session Configuration</h5>
+% if configuration:
+	<p>Using `ISessionFactory`: <code>${configuration}</code></p>
```
Using backticks has no effect on formatting. I'd prefer either their removal throughout or use of `<code>`. Also `<pre>` is not a good option because its default display is a block, not inline.

stevepiercy commented on src/pyramid_debugtoolbar/panels/templates/session.dbtmako:
```diff
@@ -0,0 +1,90 @@
+## Panel Active
+% if is_active:
+	<p>This panel <span class="label label-success">was active</span> for this
+	request, and will always display the "Session" data -- even if `.session`
```
I prefer short sentences for easier comprehension.

```suggestion
	request. It will always display the session data, even if <code>.session</code>
```

stevepiercy replied:
Also apply to `else` clause below.

stevepiercy commented on src/pyramid_debugtoolbar/panels/templates/session.dbtmako:
```diff
@@ -0,0 +1,90 @@
+	<p>This panel <span class="label label-success">was active</span> for this
+	request, and will always display the "Session" data -- even if `.session`
+	was not accessed during the request.
+	This behavior can be toggled with the Panel Settings.</p>
```
Move this outside the `if/else` and put into a separate paragraph to avoid repeating yourself and to make it more prominent.

Also I am not clear where one would access Panel Settings. Do you mean Settings > Panel Settings, in config, both, or something else?

jvanasco replied:
I've changed the text to read: "This behavior can be toggled via the Toolbar Settings tab."

The phrase "Toolbar Settings tab" has been used elsewhere in the docs.

stevepiercy commented on src/pyramid_debugtoolbar/panels/templates/session.dbtmako:
```diff
@@ -0,0 +1,90 @@
+<h5>Documentation Links</h5>
+<ul>
+	<li>API Docs:
+		<a href="http://docs.pylonsproject.org/projects/pyramid/en/latest/api/session.html">pyramid.session</a>
```
```suggestion
		<a href="https://docs.pylonsproject.org/projects/pyramid/en/latest/api/session.html">pyramid.session</a>
```

stevepiercy commented on src/pyramid_debugtoolbar/panels/templates/session.dbtmako:
```diff
@@ -0,0 +1,90 @@
+		<a href="http://docs.pylonsproject.org/projects/pyramid/en/latest/api/session.html">pyramid.session</a>
+	</li>
+	<li>Narrative Docs:
+		<a href="http://docs.pylonsproject.org/projects/pyramid/en/latest/narr/sessions.html">Session</a>
```
```suggestion
		<a href="https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/sessions.html">Session</a>
```

stevepiercy commented on src/pyramid_debugtoolbar/panels/templates/session.dbtmako:
```diff
@@ -0,0 +1,90 @@
+				<tr>
+					<th>
+						% if k in session_data["changed"]:
+							<span class="label label-info">!</span>
```
This icon alone does not tell the user what it means, and they would have to dig into this code to figure it out. Perhaps a tooltip or legend is appropriate here.

jvanasco replied:
tooltip!

stevepiercy commented on src/pyramid_debugtoolbar/templates/settings_tab.dbtmako:
```diff
@@ -5,8 +5,8 @@
-<p>Some pyramid debug toolbar panels can be activated (this enables additional features).</p>
-<p>NOTE: These settings are tracked in a cookie on a per-request basis. If you want to enable options for other requests such as via curl or a non-browser client then you should turn them on by default in your settings via the "debugtoolbar.active_panels" setting, or by setting the "pdtb_active" cookie on the client.</p>
+<p>Some pyramid_debugtoolbar panels can be activated (this enables additional features).</p>
+<p>NOTE: These settings are tracked in a cookie on a per-request basis. If you want to enable options for other requests such as via curl or a non-browser client then you should turn them on by default in your settings via the <code>debugtoolbar.active_panels</code> setting, or by setting the <code>pdtb_active</code> cookie on the client.</p>
```
👍

stevepiercy commented on src/pyramid_debugtoolbar/toolbar.py:
```diff
@@ -63,8 +63,10 @@ def __init__(
         self.visible = False
 
         # Panels can be be activated (more features) (e.g. Performace panel)
```
Would you please fix the spelling of "Performace" to "Performance"? I know it's outside this PR's scope, but I just noticed it during review.

stevepiercy commented on CHANGES.txt:
```diff
@@ -1,3 +1,33 @@
+
+- Updated toolbar javascript to better handle multiple user-activated panels.
+
+  * `split` and `join` functions now use the same delimiter.
```
reST syntax fix.

```suggestion
  * ``split`` and ``join`` functions now use the same delimiter.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+Session
+~~~~~~~
+
+Displays ingress and egress Session data if the Session was accessed during
```
```suggestion
Displays ingress and egress session data if the session was accessed during
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+Displays ingress and egress Session data if the Session was accessed during
+the request.
+
+Displays a status message indicating whether or not the Session was accessed
```
```suggestion
Displays a status message indicating whether or not the session was accessed
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+Displays a status message indicating whether or not the Session was accessed
+during the request.
+
+Advanced functionality: If the panel is enabled, the ingress and egress Session
```
```suggestion
Advanced functionality: If the panel is enabled, the ingress and egress session
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+during the request.
+
+Advanced functionality: If the panel is enabled, the ingress and egress Session
+data will always be tracked and displayed -- regardless of the Session having
```
```suggestion
data will always be tracked and displayed, regardless of the session having
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+Advanced functionality: If the panel is enabled, the ingress and egress Session
+data will always be tracked and displayed -- regardless of the Session having
+been accessed during the request. This advanced usage is offered to aid
+developers in complex debugging scenarios; most users will not need this
```
```suggestion
developers in complex debugging scenarios. Most users will not need this
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+developers in complex debugging scenarios; most users will not need this
+enabled.
+
+There are two ways to enable the extended Session display used by the
```
```suggestion
There are two ways to enable the extended session display used by the
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,6 +422,38 @@ Displays the renderings performed by Pyramid for the current page.
+#.  Under the :guilabel:`Settings` tab in the navigation bar, click the red
+    :guilabel:`X` mark. When there is a green :guilabel:`check` mark, each
+    request will have the ingress and egress data tracked and displayed on the
+    :guilabel:`Settings` panel output regardless of the Session being accessed
```
```suggestion
    :guilabel:`Settings` panel output regardless of the session being accessed
```

stevepiercy reviewed (changes requested):
This is really good work. Thank you!

I've added some change requests that are mostly formatting, a few grammar fixes, and other items to maintain UI consistency within the toolbar. I'd appreciate your consideration of them.

mmerickel commented:
What's the point of your variant of wrap_load versus using wrap_load itself? Is there a good reason this part of the code can't be more DRY?

jvanasco commented:
Thanks for the feedback. I'll address everything on my next block; likely Friday or Monday.

IIRC, I think I just didn't want to migrate wrap_load into the utils file, and have both reference it in this PR; I wanted to avoid the panels referencing each other. I can certainly do that though.

mmerickel commented:
It's fine with me if they reference each other atm. Moving it to utils would be nice too but not required.

stevepiercy reviewed (approved):
Thank you! The changes LGTM.

jvanasco commented:
@stevepiercy when you get a chance to give a spin, please let me know if you find it useful. the internal version of this had really saved me a lot of frustration debugging some problems.

mmerickel commented:
Nice work, I released 4.9 with this.

stevepiercy commented:
@jvanasco we tried this out today during my Pyramid Training for PloneConf, and the [step on sessions](https://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tutorial/sessions.html) did a good job of demonstrating the ingress and egress, while reloading the page would increase each counter. I watched participants' eyes light up, "oooh purty!" Nice work!
